File: R/forLater/YeJoo_CSVtoHTML/readHTML.R
readHTML <- function(masterHTML, csvfilename){
 
  csvFile <- parseCSV(csvfilename)
  htmlFile <- scan(file = masterHTML, what = character(0), sep = "\n", quote = "")
  newHTML <- htmlFile
 csvFile2 <- readr::read_csv(csvfilename)
  field2 <- csvFile2$params
  arr_str <- ""
  for(p in 1:length(field2)){
    if(p == 1){
      arr_str = paste(arr_str, "[", '"', field2[p], '"', sep = '')
    }
    else{
      if(p != length(field2)){
        arr_str = paste(arr_str, " ", ",", '"', field2[p], '"', sep = '')
      }
      else{
        arr_str = paste(arr_str,",", " ",'"', field2[p], '"', "]", sep = '')
      }
    }
  }
  arr_index <- grep( "other_field = ", htmlFile)
  arr_line = unlist(strsplit(htmlFile[arr_index], '='))
  newHTML[arr_index] = paste(arr_line[1], " = ", arr_str, sep = "")
  id <- csvFile$id
  field <- csvFile$field
  html_indices <- list()
  indices <- c()
  for (i in 1:length(id)){
    findstring <- paste("id=", "\"",id[i], "\"", sep = '')
    index <- grep(findstring, htmlFile)
    if (length(index) != 0){
      html_indices[[i]] <- index
      indices <- c(indices, i)
    }
  }

  for (n in 1:length(indices)){
    if (identical(id[indices[n]], "info.dephist.device.regset")){
      next
    }
    
    if (identical(id[indices[n]], "info.udm.export")){
      old_html <- unlist(strsplit(htmlFile[html_indices[[indices[n]]]], '" />'))
      old_html_end <- old_html[2]
      if(field[indices[n]] == "1"){
        old_html[2] <- "\" checked />"
      }
      else{
        old_html[2] <- "\" />"
      }
      old_html[3] <- old_html_end
      newHTML[html_indices[[indices[n]]]] <- paste(old_html[1],old_html[2],old_html[3], sep = '')
    }
    
    if (!identical(id[indices[n]], "info.dephist.device.tzone")){
      old_html <- unlist(strsplit(htmlFile[html_indices[[indices[n]]]],'value = \"\"'))
      if (length(unlist(strsplit(htmlFile[html_indices[[indices[n]]]],'value = \"\"'))) != 1){
        old_html_end <- old_html[2]
        old_html[2] <- paste("value=", "\"", field[indices[n]], "\"", ' ', sep = '')
        old_html[3] <- old_html_end
        newHTML[html_indices[[indices[n]]]] <- paste(old_html[1],old_html[2],old_html[3])
      }
    } 
  
    if (identical(id[indices[n]], "info.dephist.device.tzone")){
      f_str = paste("value=", "\"",field[indices[n]],"\"", sep = '')
      tzone_indices <- grep(f_str, htmlFile)
      if(length(tzone_indices) != 0){
        old_html <- unlist(strsplit( htmlFile[tzone_indices[1]],'">'))
        old_html_end <- old_html[2]
        old_html[2] <- paste("\"", "selected >", sep = '')
        old_html[3] <- old_html_end
        newHTML[tzone_indices[1]] <- paste(old_html[1],old_html[2],old_html[3])
      }
    }
  }
  # Write cell A into txt
  fileConn<-file("dynamic_tagmetadata.html")
  writeLines(newHTML, fileConn)
  close(fileConn)
}


parseCSV<-function(csvfilename){
  a1 <- readr::read_csv(csvfilename)
  
  local_index <- grep("info.dephist.deploy.locality", a1$field)
  local_id0 <- paste(a1[local_index, 1],'0', sep = '')
  local_id1 <- paste(a1[local_index,1],'1', sep = '')
  local_char_vector <- unlist(strsplit(as.character(a1[local_index, 3]), "[,]"))
  local_field0 <- substring(local_char_vector[1],3)
  local_field1 <- substring(local_char_vector[2],2)
  local_field1 <- substring(local_field1,1, nchar(local_field1)-2)
  local_row0 <- c(local_id0, '1', local_field0)
  local_row1 <- c(local_id1, '1', local_field1)
  a1 <- a1[-local_index,]
  a1 <- rbind(a1[1:(local_index-1),],local_row0, a1[-(1:(local_index-1)),])
  a1 <- rbind(a1[1:local_index,],local_row1, a1[-(1:local_index),])
  deploy_date_index <- grep("info.dephist.deploy.datetime.start", a1$field)
  deploy_date_id0 <- paste(a1[deploy_date_index, 1],'0', sep = '')
  deploy_date_id1 <- paste(a1[deploy_date_index, 1],'1', sep = '')
  deploy_date_vector <- unlist(strsplit(as.character(a1[deploy_date_index, 3]), " "))
  deploy_date_field0 <- deploy_date_vector[1]
  deploy_date_field1 <- deploy_date_vector[2]
  deploy_date_row0 <- c(deploy_date_id0, '1', deploy_date_field0)
  deploy_date_row1 <- c(deploy_date_id1, '1', deploy_date_field1)
  a1 <- a1[-deploy_date_index,]
  a1 <- rbind(a1[1:(deploy_date_index-1),],deploy_date_row0, a1[-(1:(deploy_date_index-1)),])
  a1 <- rbind(a1[1:deploy_date_index,],deploy_date_row1, a1[-(1:deploy_date_index),])
  
  device_date_index <- grep("info.dephist.device.datetime.start", a1$field)
  device_date_id0 <- paste(a1[device_date_index, 1],'0', sep = '')
  device_date_id1 <- paste(a1[device_date_index, 1],'1', sep = '')
  device_date_vector <- unlist(strsplit(as.character(a1[device_date_index, 3]), " "))
  device_date_field0 <- device_date_vector[1]
  device_date_field1 <- device_date_vector[2]
  device_date_row0 <- c(device_date_id0, '1', device_date_field0)
  device_date_row1 <- c(device_date_id1, '1', device_date_field1)
  a1 <- a1[-device_date_index,]
  a1 <- rbind(a1[1:(device_date_index-1),],device_date_row0, a1[-(1:(device_date_index-1)),])
  a1 <- rbind(a1[1:device_date_index,],device_date_row1, a1[-(1:device_date_index),])
  
  return(list(id=a1$field, field=a1$params))
} 

File: matlab/yejoo_preloadcsvtohtml/readHTML.m
function readHTML(masterHTML, csvfilename)


    [id2, fields2] = parseCSV2(csvfilename);
    htmlID = fopen(masterHTML);
    s = textscan(htmlID,'%s','Delimiter','\n');
    s = s{1};
    newS = s;
    fclose(htmlID);
    for p = 1:length(s)
        findstring = strcat("other_field = ");
        array_index = find(contains(s, findstring));
    end
    [field_split, unused] = strsplit(s{array_index}, "[");
    array_string = "";
    for u = 1:length(fields2)
        if u == 1,
            array_string = strcat("[", fields2{u});
        else
            if u ~= length(fields2),
                array_string = strcat(array_string, " ", ",",  fields2{u} );
            else
                 array_string = strcat(array_string, " ", ",", fields2{u}, "]");
            end
        end
    end
    [id,fields] = parseCSV(csvfilename);
    newS{array_index} = strcat(field_split{1}, " ", array_string);
    c = containers.Map;
    indices = [];
    for i = 1:length(id)
        findstring = strcat('id=',id(i));
        index = find(contains(s, findstring));
        if ~isempty(index)
            c(char(id(i))) = index;
            indices = [indices; i];
        end
    end


    for n = 1:length(indices)
        if strcmp(id(indices(n)), '"info.dephist.device.regset"')
            continue;
        end
        if strcmp(id(indices(n)), '"info.udm.export"')
            [C,matches] = strsplit(s{c(char(id(indices(n))))}, '" />');
            tempcell = C{2};
            if fields{indices(n)} == '1'
                C{2} = '" checked />';
            else
                C{2} = '" />';
            end
            C{3} = tempcell;
            newS{c(char(id(indices(n))))} = [C{1} C{2} C{3}];
        end
        if ~strcmp(id(indices(n)), '"info.dephist.device.tzone"') 
            [C,matches] = strsplit( s{c(char(id(indices(n))))},'value = ""');
            if ~isempty(matches)
                tempcell = C{2};
                C{2} = [' value=' fields{indices(n)} ' '];
                C{3} = tempcell;
                newS{c(char(id(indices(n))))} = [C{1} C{2} C{3}];
            end
        end
        if strcmp(id(indices(n)), '"info.dephist.device.tzone"')
            f_str = strcat("value=",'"',fields{indices(n)},'"');
            for m = (c(char(id(indices(n))))+ 1): (c(char(id(indices(n))))+82)
                if ~isempty(strfind(s{m}, f_str))
                     [C,matches] = strsplit( s{m},'>(');
                     tempcell = C{2};
                      C{2} = " selected >(";
                      C{3} = tempcell;
                      newS{m} = strcat(C{1},C{2},C{3});
                    break;
                end
            end
        end
    end
    % Write cell A into txt
    new_html_id = fopen('dynamic_tagmetadata.html', 'w');
    for i = 1:numel(newS)
       fprintf(new_html_id,'%s\n', newS{i});
    end
    fclose(new_html_id);
end



function [id, ret_field] = parseCSV(csvfilename)    
    ret_field = {};
    id = {};
    fid = fopen(csvfilename, 'r');
    tline = fgetl(fid);
    while ischar(tline),
        [token,remain1] = strtok(tline, ',');
         if strcmp(token,'"info.dephist.device.datetime.start"') || strcmp(token,'"info.dephist.deploy.datetime.start"') || strcmp(token,'"info.dephist.deploy.locality"')
            change_token = token(1:end-1);
            token0 = strcat(change_token, '0"');
            id{end+1} = token0;
            token1 = strcat(change_token, '1"');
            id{end+1} = token1;
         else
             id{end+1} = token;
         end  
         [remain_token, field] = strtok(remain1, ',');
         field = field(2:end);
        if strcmp(token,'"info.dephist.device.datetime.start"') || strcmp(token,'"info.dephist.deploy.datetime.start"') || strcmp(token,'"info.dephist.deploy.locality"')
            if strcmp(token, '"info.dephist.device.datetime.start"') || strcmp(token, '"info.dephist.deploy.datetime.start"')
                date_time = strsplit(field);
                ret_field{end+1} = strcat(date_time{1}, '"');
                ret_field{end+1} =  strcat('"',date_time{2});
            end
            if strcmp(token,'"info.dephist.deploy.locality"'),
                field = field(4:end-3);
                locality = strsplit(field, ', ');
                ret_field{end+1} = strcat('"',locality{1}, '"');
                ret_field{end+1} = strcat('"',locality{2}, '"');
            end
         else
             ret_field{end+1} = field;
         end   
        tline = fgetl(fid);
    end
    id = id(2:end);
    ret_field = ret_field(2:end);
    fclose(fid);
end
function [id2, ret_field2] = parseCSV2(csvfilename)
    ret_field2 = {};
    id2 = {};
    fid = fopen(csvfilename, 'r');
    tline = fgetl(fid);
    while ischar(tline)
        [token,remain1] = strtok(tline, ',');
        id2{end+1} = token;
        [remain_token, field2] = strtok(remain1, ',');
         field2 = field2(2:end);
        ret_field2{end+1} = field2;
         tline = fgetl(fid);
    end
    id2 = id2(2:end);
    ret_field2 = ret_field2(2:end);
    fclose(fid);
end
    
  